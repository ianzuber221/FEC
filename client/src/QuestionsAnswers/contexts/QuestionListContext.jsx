import React, { useState, useContext } from 'react';

const QuestionListContext = React.createContext();
const UpdateQuestionListContext = React.createContext();

export function useQuestionList() {
  return useContext(QuestionListContext);
}

export function useUpdateQuestionList() {
  return useContext(UpdateQuestionListContext);
}

export function QuestionListProvider({ children }) {
  const [questionList, setQuestionList] = useState([]);

  function updateQuestionList(newQuestionList) {
    setQuestionList(newQuestionList);
  }

  return (
    <QuestionListContext.Provider value={questionList}>
      <UpdateQuestionListContext.Provider value={updateQuestionList}>
        { children }
      </UpdateQuestionListContext.Provider>
    </QuestionListContext.Provider>
  );
}
