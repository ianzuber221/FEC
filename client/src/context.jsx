import React, { useContext, useState } from 'react';

const CurrentProductContext = React.createContext();
const UpdateCurrentProductContext = React.createContext();

export function useCurrentProductContext() {
  return useContext(CurrentProductContext);
}

export function useCurrentProductUpdate(obj) {
  return useContext(UpdateCurrentProductContext);
}
export function CurrentProductProvider({ children }) {
  const [currentProduct, setCurrentProduct] = useState('37314');

  function changeCurrentProduct(newProduct) {
    setCurrentProduct(newProduct);
  }

  return (
    <CurrentProductContext.Provider value={currentProduct}>
      <UpdateCurrentProductContext.Provider value={changeCurrentProduct}>
        {children}
      </UpdateCurrentProductContext.Provider>
    </CurrentProductContext.Provider>
  );
}