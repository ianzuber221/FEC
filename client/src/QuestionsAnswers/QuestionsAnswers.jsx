import React, { useState, useEffect, forwardRef } from 'react';
import styled from 'styled-components';
import _ from 'underscore';
import { getQuestions, getProductInfo } from './lib/api/githubAPI';
import Search from './SearchBar/Search';
import QuestionList from './QuestionList/QuestionList/QuestionList';
import MoreQuestions from './OptionButtons/MoreQuestions';
import AddQuestion from './OptionButtons/AddQuestion';
import { useCurrentProductContext } from '../context';
import { useQuestionList, useUpdateQuestionList } from './contexts/QuestionListContext';

const QuestionAnswers = forwardRef((props, ref) => {
  const productId = useCurrentProductContext();
  const questionList = useQuestionList();
  const setQuestionList = useUpdateQuestionList();
  const [filteredKeyword, setFilteredKeyword] = useState('');
  const [productInfo, setProductInfo] = useState({});
  const [expanded, setExpanded] = useState(false);
  const [hasMore, setHasMore] = useState(true);
  // const [count, setCount] = useState(200);
  const [page, setPage] = useState(1);
  const count = 10;

  const renderQuestions = (init) => {
    if (init) {
      setPage(1);
      setHasMore(true);
    }
    const targetPage = init ? 1 : page;
    getQuestions(productId, targetPage, count)
      .then((result) => {
        if (result.data.results.length === 0) {
          setHasMore(false);
          return;
        }
        if (targetPage === 1) {
          setQuestionList(sortQuestions(result.data.results));
        } else {
          const newQList = {};
          questionList.concat(result.data.results).forEach((item) => {
            newQList[item.question_id] = item;
          });
          const sorted = sortQuestions(Object.values(newQList));
          setQuestionList(sorted);
        }
      });
  };

  useEffect(() => {
    renderQuestions(true);
    getProductInfo(productId)
      .then((result) => {
        setProductInfo(result.data);
      });
  }, [productId]);

  useEffect(() => {
    renderQuestions(true);
  }, [expanded]);

  return (
    <DivContainer id="question-and-answers" className="main-widget-container" ref={ref}>
      <div className="main-widget-title">QUESTIONS & ANSWERS</div>
      <Search setFilter={setFilteredKeyword} />
      <QuestionList
        renderQuestions={renderQuestions}
        keyword={filteredKeyword}
        productName={productInfo.name}
        expanded={expanded}
        page={page}
        setPage={setPage}
        hasMore={hasMore}
      />
      <ButtonContainer>
        <MoreQuestions
          expanded={expanded}
          setExpanded={setExpanded}
          Button={Button}
        />
        <AddQuestion
          renderQuestions={renderQuestions}
          Button={Button}
          productName={productInfo.name}
          setHasMore={setHasMore}
        />
      </ButtonContainer>
    </DivContainer>
  );
})

export default QuestionAnswers;

const sortQuestions = (questions) => (
  _(questions).chain()
    .sortBy('question_date')
    .reverse()
    .sortBy('question_helpfulness')
    .reverse()
    .value()
);

const DivContainer = styled.div`
  display: flex;
  flex-direction: column;
  width: 70%;
  color : #3d3c3c;
  font-size: 15px;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Ubuntu, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif;
  border-bottom: 0.5px solid black;
`;

const Button = styled.button`
  padding: 15px;
  border: 1px solid #3d3c3c;
  background-color: white;
  margin-right: 10px;
  font-weight:500px!important;
  &:hover {
    cursor: pointer;
    opacity: 60%;
  }
`;

const ButtonContainer = styled.div`
  display: flex;
  margin-top: 30px;
`;
