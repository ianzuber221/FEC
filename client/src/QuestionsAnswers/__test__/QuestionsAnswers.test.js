/**
 * @jest-environment jsdom
 */
/* eslint-disable react/jsx-filename-extension */

import React from 'react';
import { render, screen, waitForElementToBeRemoved } from '@testing-library/react';
import QuestionsAnswers from '../QuestionsAnswers';
import '@testing-library/jest-dom/extend-expect';
import { CurrentProductProvider } from '../../context';
import { QuestionListProvider } from '../contexts/QuestionListContext';

require('dotenv').config();

it('matches snapshot', async () => {
  const { getAllByText } = await render(
    <CurrentProductProvider>
      <QuestionListProvider>
        <QuestionsAnswers />
      </QuestionListProvider>
    </CurrentProductProvider>
  );

  await waitForElementToBeRemoved(screen.getByText('No Questions Available.'));
  expect(getAllByText('Q:').length > 0).toBe(true);
});