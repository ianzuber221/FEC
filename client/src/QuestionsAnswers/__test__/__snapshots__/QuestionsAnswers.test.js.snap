// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-efBctP knyPhM"
    id="question-and-answers"
  >
    <div
      class="sc-dPyBCJ dxRBDO"
    >
      QUESTIONS & ANSWERS
    </div>
    <form
      class="sc-bczRLJ brtfGI"
    >
      <div
        class="sc-hKMtZM giyKxL"
      >
        <input
          class="sc-gsnTZi bvoSbL"
          placeholder="HAVE A QUESTION? SEARCH FOR ANSWERS..."
          type="text"
          value=""
        />
        <svg
          aria-hidden="true"
          class="svg-inline--fa fa-magnifying-glass sc-dkzDqf deXTCe"
          data-icon="magnifying-glass"
          data-prefix="fas"
          focusable="false"
          role="img"
          viewBox="0 0 512 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M500.3 443.7l-119.7-119.7c27.22-40.41 40.65-90.9 33.46-144.7C401.8 87.79 326.8 13.32 235.2 1.723C99.01-15.51-15.51 99.01 1.724 235.2c11.6 91.64 86.08 166.7 177.6 178.9c53.8 7.189 104.3-6.236 144.7-33.46l119.7 119.7c15.62 15.62 40.95 15.62 56.57 0C515.9 484.7 515.9 459.3 500.3 443.7zM79.1 208c0-70.58 57.42-128 128-128s128 57.42 128 128c0 70.58-57.42 128-128 128S79.1 278.6 79.1 208z"
            fill="currentColor"
          />
        </svg>
      </div>
    </form>
    <div
      id="question-list"
    >
      <div
        class="sc-himrzO kJLYaB"
      >
        <div
          class="individual-question"
        >
          <div
            class="sc-bZkfAO juhrtI"
          >
            <div
              class="sc-ikZpkk jCYdMS"
            >
              <span
                class="sc-kLLXSd cpAjvV"
              >
                Q:
              </span>
              <b
                class="sc-jIZahH hnxZwi"
              >
                what is MY QUESTION?
              </b>
            </div>
            <div
              class="css-1iivxpu-MuiStack-root"
            >
              <div
                class="sc-gKXOVf ggfvoY helpful"
              >
                <span
                  class="sc-eCYdqJ edpWWz"
                >
                  Helpful?
                </span>
                <u
                  class="sc-jSMfEi jlKawR"
                  role="button"
                  tabindex="-1"
                >
                  Yes
                </u>
                <span>
                  (1)
                </span>
              </div>
              <hr
                class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
              />
              <u
                role="button"
                tabindex="-1"
              >
                Add Answer
              </u>
            </div>
          </div>
          <div
            class="sc-iIPllB eZDcGx"
          >
            <span
              class="sc-kLLXSd cpAjvV"
            >
              A:
            </span>
            <div
              class="sc-ezWOiH gvvbdh"
            >
              <div
                class="sc-gicCDI pqQNY"
                id="scrollableDiv"
              >
                <div
                  class="sc-ivTmOn cRtkON answer"
                >
                  <div
                    class="answer-text"
                  >
                    testing images
                  </div>
                  <div
                    class="sc-cxabCf jGAeqJ"
                  >
                    <img
                      class="sc-dmRaPn gTeHXh"
                      src="http://res.cloudinary.com/dl9zxpaoq/image/upload/v1658255367/cv9ghenayxntg97bh4cp.jpg"
                    />
                    <img
                      class="sc-dmRaPn gTeHXh"
                      src="http://res.cloudinary.com/dl9zxpaoq/image/upload/v1658255371/b1vqersvcoabfd6zbvzd.png"
                    />
                  </div>
                  <div
                    class="css-1iivxpu-MuiStack-root"
                  >
                    <span>
                      <span>
                        by 
                        <b>
                          Seller
                        </b>
                      </span>
                      , July 18, 2022
                    </span>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
                    />
                    <div
                      class="sc-gKXOVf ggfvoY helpful"
                    >
                      <span
                        class="sc-eCYdqJ edpWWz"
                      >
                        Helpful?
                      </span>
                      <u
                        class="sc-jSMfEi jlKawR"
                        role="button"
                        tabindex="-1"
                      >
                        Yes
                      </u>
                      <span>
                        (0)
                      </span>
                    </div>
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
                    />
                    <span>
                      <u
                        role="button"
                        tabindex="-1"
                      >
                        Report
                      </u>
                    </span>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="individual-question"
        >
          <div
            class="sc-bZkfAO juhrtI"
          >
            <div
              class="sc-ikZpkk jCYdMS"
            >
              <span
                class="sc-kLLXSd cpAjvV"
              >
                Q:
              </span>
              <b
                class="sc-jIZahH hnxZwi"
              >
                Test question
              </b>
            </div>
            <div
              class="css-1iivxpu-MuiStack-root"
            >
              <div
                class="sc-gKXOVf ggfvoY helpful"
              >
                <span
                  class="sc-eCYdqJ edpWWz"
                >
                  Helpful?
                </span>
                <u
                  class="sc-jSMfEi jlKawR"
                  role="button"
                  tabindex="-1"
                >
                  Yes
                </u>
                <span>
                  (0)
                </span>
              </div>
              <hr
                class="MuiDivider-root MuiDivider-fullWidth MuiDivider-vertical MuiDivider-flexItem css-18i4s2i-MuiDivider-root"
              />
              <u
                role="button"
                tabindex="-1"
              >
                Add Answer
              </u>
            </div>
          </div>
          <div
            class="sc-iIPllB eZDcGx"
          >
            <span
              class="sc-kLLXSd cpAjvV"
            />
            <div>
              No answers.
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="sc-jOrMOR ibBxDu"
    >
      <div>
        <button
          class="sc-cTQhss iUUzGc"
          data-testid="more-question-btn"
          type="submit"
        >
          MORE ANSWERED QUESTIONS
        </button>
      </div>
      <div
        id="add-question-btn"
      >
        <button
          class="sc-cTQhss iUUzGc"
          data-testid="add-question-btn"
          type="submit"
        >
          ADD A QUESTION +
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
